torch_xla/csrc/helpers.cpp

```diff
@@ -624,13 +680,20 @@ xla::Shape XlaHelpers::GetPromotedDynamicShape(const xla::Shape& shape1,
       PromoteType(shape1.element_type(), shape2.element_type()), upper_bounds,
       dyn_dims);
 
+  std::cout << xla::ShapeUtil::HumanString(promoted_shape);
   return promoted_shape;
 }
 
-std::pair<xla::XlaOp, xla::XlaOp> XlaHelpers::ValidateShapes(xla::XlaOp op1,
-                                                             xla::XlaOp op2) {
+std::pair<xla::XlaOp, xla::XlaOp> XlaHelpers::PromoteShapes(xla::XlaOp op1,
+                                                            xla::XlaOp op2) {
   const xla::Shape& shape1 = ShapeHelper::ShapeOfXlaOp(op1);
   const xla::Shape& shape2 = ShapeHelper::ShapeOfXlaOp(op2);
+
+  xla::Shape shape = GetPromotedShape(shape1, shape2);
+  if (shape.is_unbounded_dynamic()) {
+    return ImplicitBroadcastWithUnboundedDynamicShapes(op1, op2, shape);
+  }
+
   if (xla::ShapeUtil::Compatible(shape1, shape2)) {
     // Fast path shortcut if the shapes already matches in dimensions.
     return std::pair<xla::XlaOp, xla::XlaOp>(op1, op2);
```

Comment:
I agree. What the back-to-back if statements says that:
// Ignore fp-precison if the shapes of op1 and op2 have same dimensions, otherwise the element-types must
// exactly match. 

In other words shape promotion is allowed only when the element-types are the same.

Reply:
cc @qihqi

Reply:
It looks like `Compatible` does enforce fp-precision, but allows unbounded shape mismatches. This means that `SameElementType` is a looser check than Compatible. Perhaps we can remove the `Compatible` check all together.

Reply:
Yup, that makes sense. done.